오픈마켓 [STEP 1] 산체스, 예스코치

YesCoach opened the pull request:
# 오픈마켓 STEP 1 PR
@YebinKim 
븨븨 안녕하세요!! STEP 1-1,2 를 어느정도 구현해서 이렇게 PR를 보냅니다.
네트워크 통신과 관련해서 어려움이 많았었는데 븨븨 덕분에 차분하게 잘 구현할 수 있었습니다.

## 고민한 점
1. Asset에 mock 데이터들을 넣어서 테스트에 활용해보고자 하였는데 쉽지 않아서 일단은 UnitTest 내부에서 직접 테스트 데이터를 만들어서 진행했습니다. UnitTest에서 사용할 테스트 데이터를 이렇게 직접 만들어서 써도 무방한지, Asset에 넣어서 이를 사용하는 방법을 더 고민해봐야할지 궁금합니다. 
2. multipart 폼을 구성해서 post나 patch 등을 수행하고 있는데, 내부의 기능들을 분리해서 구현을 하는게 좋을지 궁금합니다... 가장 눈에 띄는 점은 이미지(Media 타입)를 올릴때 매개변수로 전달해주어야 하는데 지금은 타입 내부에서 하드코딩으로 직접 주고 있습니다. 저희가 코드의 분리에 있어서 좀 더 생각해볼 점이 있다면 알려주시면 감사드립니다!!

이외에도 네트워크 통신이나 테스트 코드와 관련해서 부족한 부분이 있다면 많이많이 지적해주세요ㅠㅠ 잘부탁드립니다!!

sanches37 committed:
feat: item, items 타입 구현

YesCoach committed:
feat: 서버에서 데이터를 받는 NetworkManager 타입 구현

* ItemData와 ItemsData를 공용으로 사용하기 위해 구조체 수정

sanches37 committed:
feat: 테스트용 JsonDecoder 타입 구현 및 Asset 추가
Decoder, MockDecoder protocol 타입 구현
테스트를 위한 MockDataAsset 추가

YesCoach committed:
test: Mock Data를 Decode 하는 UnitTest 구현

sanches37 committed:
refactor: JsonDecoder 타입 내 메서드 수정
JsonDecoder 타입 내 decodedJseonFromDataAsset 메서드 삭제
NetworkManager 내 decode 처리를 JsonDecoder.decodedJsonFromData 로 구현

YesCoach committed:
refactor: Media 타입 및 generateBoundary 메서드 생성
multipart에 사용되는 Media 타입 구현.
NetworkManager에서 generateBoundary 메서드 구현.

sanches37 committed:
feat: postItem 메소드 구현
멀티파트 형식으로 서버에 POST 를 요청하기 위해
createDataBody 메소드 구현
PostItemData 타입 구현

YesCoach committed:
refactor: url과 method를 다루는 enum 타입 ApiForm 생성
getItems에 page 매개변수를 받도록 수정

sanches37 committed:
feat: PATCH, DELETE 기능 추가
ietm에 GET으로 접근하는 getItem 메서드 추가

YesCoach committed:
refactor: ApiFormat 타입의 path 메소드의 이름 변경
의미에 좀더 부합한 url로 이름 변경

YebinKim commented on OpenMarket/OpenMarket/Network/ApiFormat.swift:
```diff
@@ -0,0 +1,40 @@
+
+import Foundation
+
+enum ApiFormat {
```
사용하는 API를 enum 타입으로 편리하게 사용할 수 있게 작성하셨군요!
멋진 활용입니다 😆👍

sanches37 replied:
감사합니다 ~ 😀

YebinKim commented on OpenMarket/OpenMarket/Network/ApiFormat.swift:
```diff
@@ -0,0 +1,40 @@
+        case .getItems:
+            return Self.baseUrl + "items/"
+        case .post, .getItem, .patch, .delete:
+            return Self.baseUrl + "item/"
```
현재는 url을 사용할 때(네트워크 로직에서) 각 아이템 또는 페이지 번호를 url String 에 추가하는 구조로 보이는데요!
enum의 연관값(Enum associated values)을 활용하면 분기문 안에서 각각의 아이템/페이지 정보를 추가할 수도 있습니다 😊

YesCoach replied:
```swift
enum APIFormat {
    case getItems(page: Int)
    case getItem(id: Int)
    case post
    case patch(id: Int)
    case delete(id: Int)

    var path: String {
        switch self {
        case .getItems(let page):
            return "items/" + String(page)
        case .post:
            return "item"
        case .getItem(let id), .patch(let id), .delete(let id):
            return "item/" + String(id)
        }
    }
```
이렇게 사용해봤습니다!! 
리팩토링 과정에서 다시 변경하느라 연관값 사용을 결과론적으로는 안했지만, 기능과 장점에 대해서 많이 배울 수 있었습니다ㅎㅎ
추가로 Resert 타입도 잘 몰랐는데 enum에 연관값을 사용하더라고요.. 오늘 처음 알았는데 덕분에 이해하기 쉬웠습니다👍👍

YebinKim commented on OpenMarket/OpenMarket/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,188 @@
+//
+
+import Foundation
+typealias parameters = [String: Any]
```
파라미터 타입에 대한 선언이 소문자 카멜케이스로 되어 있는 걸로 보이네요!
Swift API Guidelines 에서 타입 및 프로토콜 요소는 대문자 카멜케이스의 사용을 권장하고 있습니다 ☺️

https://swift.org/documentation/api-design-guidelines/#general-conventions

sanches37 replied:
감사합니다 . Parameters 로 적용했습니다. 😄

YebinKim commented on OpenMarket/OpenMarket/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,188 @@
+        session.dataTask(with: request) { data, response, error in
+            if let error = error {
+                print(error.localizedDescription)
+            }
```
에러가 있어도 dataTask의 컴플리션 블럭을 탈출하지 않고 아래의 로직을 실행할 것으로 보이는데요.
혹시 이건 의도된 설계일까요? 🤔

sanches37 replied:
전체적으로 리팩토링을 하며 수정했습니다~ 😆

YesCoach replied:
음... 저희가 놓친 부분인거같습니다😅 탈출하는 로직으로 수정했습니다!!

YebinKim commented on OpenMarket/OpenMarket/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,188 @@
+                print(error.localizedDescription)
+            }
+            guard let response = response else { return }
+            print(response)
```
print 메서드를 로그로 사용하고자 한다면 어떤 메서드의 로그인지를 함께 출력해주는 것이 더 로그 파악에 도움될 거 같아요~
매번 써주는 것이 번거롭다면 스위프트에서 기본으로 제공하는 로그 타입이나 `debugPrint(_:separator:terminator:)` 을 활용해보는 방법 또한 있습니다! (가능하면 머지 커밋에는 print 메서드를 남기지 않는 것이 좋습니다)

sanches37 replied:
debugPrint 를 사용하였습니다. 다음 단계에서 error enum 타입 안에 var 로 호출을 추가해보도록 하겠습니다.

YebinKim commented on OpenMarket/OpenMarket/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,188 @@
+                print(error.localizedDescription)
+            }
+            guard let response = response else { return }
+            print(response)
```
response 는 dataTask의 컴플리션 블럭 안에서 사용하지 않고 있는 것으로 보이네요!
response 값을 캡처하고 검증하는 이유가 있을까요? 🤔

sanches37 replied:
생각없이 사용한 것 같습니다. 전체적으로 리팩토링하며 수정했습니다~ 😊

YesCoach replied:
이부분도 저희가 대수롭지 않게 생각했었는데 맞네요!!
response가 담고 있는 status code를 체크하는 로직을 추가했습니다..! 200번대의 성공 상태코드가 담겨있는지 체크해보았어요

YebinKim commented on OpenMarket/OpenMarket/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,188 @@
+            
+            if let error = error {
+                print(error.localizedDescription)
+            }
```
이전에 회의실에서도 한 번 얘기 나누었던 거 같은데 응답을 확인하기 전 에러를 먼저 체크하는 게 필요할 거 같아요~

dataTask 로직의 작성이 각 메서드마다 통일되어 있지 않은 것으로 보이네요 🤔
**if 구문과 guard 구문 중 어떤 것을 사용해야 할지**,
**어떤 값에 대한 검증이 먼저 이루어져야 할지**
두 가지를 중심으로 생각하며 리팩토링을 진행해보면 좋을 거 같아요!

sanches37 replied:
말씀하신 부분 리팩토링 했습니다. error 부분은 else 부분을 표시할 필요가 없어서 if let 을 사용했습니다 😊

YebinKim commented on OpenMarket/OpenMarket/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,188 @@
+        if let parameters = params {
+            for (key, value) in parameters {
+                body.append("--\(boundary)\r\n")
+                body.append("Content-Disposition: form-data; name=\"\(key)\"\r\n\r\n")
```
반복되는 개행 `\r\n` 은 로컬 프로퍼티로 정의하면 가독성이 더욱 좋아질 거 같아요~

sanches37 replied:
lineBreakePoint 상수에 담아 사용했습니다. 감사합니다 😌

YebinKim reviewed (changes requested):
예스코치, 산체스 안녕하세요~
3주간 오픈마켓 프로젝트 리뷰를 함께할 비비입니다 😄
스텝1에서 모델 타입, 네트워킹 타입 작성은 물론 multipart/form 포맷까지 활용하느라 고생 많으셨어요!
커밋 내역을 보니 페어프로그래밍 또한 균형있게 역할을 나눠서 잘 진행하신 거 같네요 🤩👏
그럼 PR 질문부터 답변 드릴게요~!

---

### 1. UnitTest에서 사용할 테스트 데이터를 테스트 코드 안에서 직접 생성하는 것과 Asset에 넣어서 활용하는 방법에 관한 질문

테스트 데이터를 테스트 코드 안에서 직접 생성하는 것에 문제가 있는 것은 아니지만,
Asset 에 있는 json 데이터를 불러와 테스트 데이터를 생성할 수 있다면 더 좋을 거 같네요!

구체적으로 어떤 부분에 어려움이 있었는지가 궁금한데요 🤔
번들의 파일을 가져오지 않고 Asset 을 통해 반드시 가져오려고 하는 이유도 있을까요?? 



---

### 2. multipart/form 을 활용한 API 메서드 로직 개선에 관한 질문

타입 내부에서 하드코딩을 하고 있다는 것이 `PatchItemData` 과 `PostItemData` 의 `parameter()` 메서드를 말씀하시는 게 맞을까요? 🤔
이 부분과 관련해서는 `Encodable` 프로토콜을 활용해 타입을 json 으로 인코딩하는 방법으로 개선해볼 수 있을 거 같습니다! 😊

코드의 분리에 대해서는 메서드의 실제 활용 방법을 보며 이후 스텝에서 좀 더 코멘트 드릴 수 있도록 할게요~
현재는 메서드 안에 **테스트 로직**과 **사용하지 않는 응답값**이 섞인 부분과 **if구문, guard 구문 활용**을 고민하며 리팩토링을 진행하고
이후 프로젝트를 진행하며 네트워킹 결과 데이터의 비동기 활용에 대해 코드를 이어서 작성해보면 분리의 필요성이 더욱 보일 거 같습니다!



---

Step1 진행 고생 많으셨어요! 😆👏
전체적으로 설계와 Swift 문법의 활용이 좋아서 자잘한 부분에 대해서만 코멘트 드렸는데요.
DM방에서 말씀드렸듯 이번 스텝의 핵심 경험 중 `네트워크 상황과 무관한 네트워킹 데이터 타입의 단위 테스트(Unit Test)` 가 수행되지 않은 거 같아 Request Changes 로 코멘트 드릴게요~

<img width="509" alt="핵심경험" src="https://user-images.githubusercontent.com/38066039/129733017-354a839e-c175-4dfe-94a3-cad3374e02b3.png">

네트워킹 타입 테스트 진행까지 힘내봅시다!! 💪🔥

sanches37 committed:
feat: URLSessionProtocol,MockURLSessionDataTask 타입구현
목객체 구현을 위한 URLSessionProtocol 프로토콜 구현
resume을 재정의하는 MockURLSessionDataTask 클래스 구현
typealias parameter -> Parameter 로 네임 변경

YesCoach committed:
refactor: NSAsset을 사용하는 Unit Test로 수정
MockJsonDecoder 타입 생성

sanches37 committed:
feat: MockURLSession 타입 구현, 의존성 주입
NetworkManager 타입에 session 에 대한 의존성 주입 구현
dataTask 내 로직 수정

YesCoach committed:
test: 네트워크 무관 테스트 케이스 생성

sanches37 committed:
test

YesCoach committed:
refactor: NetworkManager 수정
* NetworkManager의 메소드들을 commuteAPI 메소드로 통합
* URLRequest 생성 로직을 메소드로 분리
* Protocol 구현 -> Requestable, RequestableWithDataBody(-WithoutDataBody)
* enum 분리 -> APIFormat을 APIURL, APIMethod로 분리
* ContentType, MimeType enum 생성

sanches37 committed:
refactor: post 작동 불량 해결
이미지 전달시 스펠링 틀린 부분 체크

sanches37 committed:
refactor: createRequest 수정, commuteWithAPI 수정
createRequest 내 Dlelte시 hpptBody 적용 부분 구현
commuteWithAPI 내 request 검사 코드 구현
DeleteAPI 로직 수정
PatchAPI 로직 수

YesCoach committed:
Refactor: MockURLSession 수정 및 테스트 코드 추가

sanches37 committed:
refactor: 테스트 코드 수정 및 /r/n을 변수로 사용
테스트코드 실패하는 부분 수정 적용
"/r/n" 을 lineBreakPoint 상수로 사용

sanches37 commented:
의존성 주입을 하면서 다른 로직도 많은 부분을 수정했습니다. request에 포함되는 url, httpMethord, parameter, httpbody 에 관한 검증을 쉽게 하기 위해 하나로 묶어서 전달하는 방식으로 코드도 수정했습니다. 비비께서 의존성 주입을 해보는게 어떤지 권유를 해주셨는데, 직접 구현해보고 필요한 부분도 리팩토링을 해보니 많은 부분을 배운 것 같습니다. 감사합니다!! 🥰😊😆

sanches37 committed:
테스트 값 get -> post 로 수정

YebinKim commented on OpenMarket/OpenMarket/Protocol/NetworkProtocol.swift:
```diff
@@ -0,0 +1,18 @@
+
+import Foundation
+
+protocol URLSessionProtocol {
+    func dataTask(with request: URLRequest, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask
+}
```
로컬/온라인 상태의 네트워크 동작을 프로토콜로 추상화 하셨군요! 🤩
의존성 주입을 위한 타입 추상화를 멋지게 구현해내셨습니다 ☺️👍

YebinKim commented on OpenMarket/OpenMarket/Protocol/Requestable.swift:
```diff
@@ -0,0 +1,21 @@
+
+import UIKit
+
+protocol Requestable {
+    var url: String { get }
+    var method: APIMethod { get }
+    var contentType: ContentType { get }
+}
+
+protocol RequestableWithoutBody: Requestable{}
+
+protocol RequestableWithBody: Requestable {
+    var parameter: [String: Any]? { get }
+    var items: [Media]? { get }
+}
```
request 데이터를 프로토콜로 선언함으로써 각각의 API를 추상화 하셨군요!
네트워킹 타입을 유연하게 구현하는 멋진 활용입니다 ☺️👏

YebinKim commented on OpenMarket/OpenMarket/Network/RequestableAPI.swift:
```diff
@@ -0,0 +1,67 @@
+    var contentType: ContentType = .json
+    
+    init(id: Int) {
+            self.url = APIURL.getItem.description + "\(id)"
```
들여쓰기가 맞지 않는 코드가 일부 보이네요 👀
들여쓰기는 수정이 필요한 코드를 드래그하고 `control + i` 단축키를 누르면 일괄 맞출 수 있어요~

sanches37 replied:
들여쓰기도 잘 확인하여 수정해보겠습니다. 감사합니다 😀

YebinKim commented on OpenMarket/OpenMarket/Network/Media.swift:
```diff
@@ -0,0 +1,47 @@
+        switch mimeType {
+        case .jpeg:
+            self.filename = "\(arc4random()).jpeg"
+            guard let data = image.jpegData(compressionQuality: 0.7) else {
+                return nil
+            }
```
image 의 jpeg 변환이 실패했을 때는 guard 구문 위의 코드도 동작할 필요가 없지 않을까요? 🤔

sanches37 replied:
코드 순서를 바꿔서 이미지 변환에 실패했을 때 다른 코드들이 동작하지 않도록 수정하겠습니다. 😃

YebinKim commented on OpenMarket/OpenMarket/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,124 @@
+                return completion(.failure(NetworkError.unownedData))
+            }
+            debugPrint(String(decoding: data, as: UTF8.self))
+            DispatchQueue.main.async {
+                completion(.success(data))
+            }
```
특별히 data 를 메인 스레드에서 전달해야 할 필요성이 지금은 보이지 않는데요.
혹시 메인 스레드에서 컴플리션 핸들러를 동작시키는 이유가 있을까요? 🤔

sanches37 replied:
스탭2를 조금 진행했을 때 컴플리션 핸들러로 넘겨주는 부분을 화면 구성에 사용되어서 main 스레드로 넘겼는데, 생각해보니 컴플리션 핸들러로 넘기는 모든 부분을 메인스레드에서 작업할 필요가 없다고 생각이 되네요. 코치와 협의 후에 수정할 수 있도록 하겠습니다! 감사합니다

YebinKim commented on OpenMarket/OpenMarket/Network/NetworkManager.swift:
```diff
@@ -0,0 +1,124 @@
+                  (200...299).contains(response.statusCode) else {
+                return completion(.failure(NetworkError.unownedResponse))
+            }
+            debugPrint(response)
```
print 메서드를 debugPrint 메서드로 변경하셨군요!
빠른 리뷰 반영 고맙습니다 ☺️

YebinKim commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,135 @@
+            XCTAssertEqual(error, JsonError.decodingFailed)
+        } catch { }
+    }
+    
+    
```
개행이 2줄 이상인 코드가 일부 보이는데요 🤔
개행은 특별한 경우가 아니라면 1줄로 사용하는 컨벤션을 사용하는 곳이 많습니다 ㅎㅎ (예를들어 [구글](https://google.github.io/swift/#vertical-whitespace) 등)
이 코멘트는 Swift의 공식 가이드라인은 아니니 참고만 부탁드려요~

sanches37 replied:
꼼꼼하게 체크하지 못한 것 같습니다 전체적으로 수정하겠습니다. 😅

YebinKim commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,135 @@
+        //given
+        let expectInputValue = "MockItems"
+        guard let jsonData = try? MockJsonDecoder.receiveDataAsset(assetName: expectInputValue) else {
+            return
```
jsonData 를 Asset에서 찾지 못 했을 때 return 만 하고 있기 때문에 잘못된 expectInputValue 를 넣어도 테스트가 성공하고 있네요!
guard 구문 검증에 실패했을 때도 테스트를 실패 시키는 코드를 추가하는게 어떨까요? 🤔

이하 다른 테스트에서도 expectInputValue 또는 expectResult 를 다르게 넣었을 때 테스트가 실패하고 있는지 확인이 필요할 거 같아요~

sanches37 replied:
테스트 부분 다시 한번 확인해서 수정해보겠습니다. 감사합니다.🙏🏼

YebinKim reviewed (approved):
산체스, 예스코치 안녕하세요~
Step1 수정까지 고생 많으셨습니다!

로컬 테스트가 가능한 네트워킹 타입을 멋지게 구현하셨군요! 😆👍
API 서버와 통신할 `NetworkManager` 타입에 **의존성 주입을 활용함으로써 유연한 설계**가 되었고,
통신에 필요한 타입들을 각각 선언함으로써 재사용성과 범용성 또한 높아졌네요 👏👏

이번 수정을 통해 Step1 에서의 핵심 경험은 모두 수행이 된 거 같아 이대로 머지 진행할게요-!
Guard 구문을 통한 빠른 종료, 메인 스레드 사용, 테스트의 결과와 관련한 코멘트 및 기타 컨벤션 관련 코멘트가 있는데 아래에서 확인하시고 필요하다면 다음 스텝에 반영 부탁드려요 ☺️🙏
